src/Processors/QueryPlan/ReadFromRemote.cpp

```diff
@@ -223,12 +246,44 @@ void ReadFromRemote::addPipe(Pipes & pipes, const ClusterProxy::IStreamFactory::
 void ReadFromRemote::initializePipeline(QueryPipelineBuilder & pipeline, const BuildQueryPipelineSettings &)
 {
     Pipes pipes;
-    for (const auto & shard : shards)
+
+    /// We have to create a pipe for each replica
+    /// FIXME: The second condition is only for tests to work, because hedged connections enabled by default.
+    if (context->getSettingsRef().max_parallel_replicas > 0 &&
+        !context->getSettingsRef().enable_sample_offset_parallel_processing &&
+        !context->getSettingsRef().use_hedged_requests)
```

Comment:
I suppose we can ignore `use_hedged_requests` for some time, if it does not work with parallel reads.